### PR TITLE
Support eager subdiagnostics again

### PR DESCRIPTION
**Pull request by mejrs**
See https://github.com/rust-lang/rust/pull/104941#discussion_r1051135746

I'm not sure how to add a test for this. But I did pick some of the diagnostic structs in the mentioned PR and it works with them.

### CODE REVIEW

File: src/test/ui-fulldeps/session-diagnostic/diagnostic-derive.rs [hunk 1]
```diff
@@ -740,14 +732,6 @@ pub(crate) struct SubdiagnosticWithSuggestion {
...
 }
 
-#[derive(Diagnostic)]
-#[diag(compiletest_example)]
-struct SubdiagnosticEagerSuggestion {
-    #[subdiagnostic(eager)]
-    //~^ ERROR `#[subdiagnostic(...)]` is not a valid attribute
-    sub: SubdiagnosticWithSuggestion,
-}
-
```
Comment (davidtwco) [inline src/test/ui-fulldeps/session-diagnostic/diagnostic-derive.rs:750]: You can just leave these structs but remove the error annotations to test this.

**Review by davidtwco — APPROVED** [review]
Apologies for the delay in reviewing, r=me after fixing the comment.

File: src/test/ui-fulldeps/session-diagnostic/diagnostic-derive.rs (hunk 1 above)
Comment (mejrs) [reply src/test/ui-fulldeps/session-diagnostic/diagnostic-derive.rs:750]: Done :)

**Comment by dtolnay** [comment]
Rebased to resolve conflict with #106734, since this was showing not mergeable in bors.
https://bors.rust-lang.org/queue/rust
@bors r=davidtwco

**Pull request merged**